### DISCUSSION TITLE
Roadmap v5

### DISCUSSION THREAD

**Discussion by darrachequesne** [socketio/socket.io#4529] [General]
Here is the list of changes for the next major version of Socket.IO. This list is obviously up for discussions, so please do not hesitate to comment here.

- [ ] remove the reference to the handshake request (`socket.request`)

Rationale: keeping a reference to the `http.IncomingMessage` increases the memory footprint. Note: we should provide an alternative to be used with `express-session` and such.

Related: https://github.com/socketio/socket.io/issues/3792

- [ ] when calling `io.close()` do not close the HTTP server if it was not created by the library

Related: https://github.com/socketio/socket.io/discussions/4002#discussioncomment-4157704

- [x] use private properties ("#") in order to improve the minification of the client bundle

Note: we might need to drop support for some older versions of Node.js.

Update: using private class fields does not significantly reduce the bundle size after minification + gzip (and is a breaking change since it is only supported starting with Node.js 14.6.0) so this change makes little sense for now.

- [x] remove the "error" event emitted by the `socket` object

Related: https://github.com/socketio/socket.io/issues/2047

Update: fixed in https://github.com/socketio/socket.io/commit/15af22fc22bc6030fcead322c106f07640336115

- [x] implement various quality of services: https://github.com/socketio/socket.io/issues/3649

Rationale: currently we have "at-most-once" semantics, possibly "at-least-once" with acknowledgements and manual retry, but we could provide additional guarantees. Might be linked to https://github.com/socketio/socket.io/discussions/4510

Update: implemented in https://github.com/socketio/socket.io-client/commit/655dce97556a1ea44a60db6b694d0cfd85b5f70f

- [ ] prevent users from shooting themselves in the foot with `socket.leave(socket.id);` 

Related: https://github.com/socketio/socket.io/issues/4524

[...]

**Comment by qiulang** [comment]
Hi what about [[RFC] Lite javascript client #4393](https://github.com/socketio/socket.io/discussions/4393)

**Reply by darrachequesne** [reply]
That's a great idea! It should be independent from v5 though.

**Comment by 655452** [comment]
That's really greate functionality

**Comment by ghost** [comment]
May I suggest to improve the usability of the RemoteSocket interface?
I fell in a lot of issues by not knowing how to properly handle a remote fetched socket... 
My proposal would be to:

- Add a documentation page for RemoteSocket explaining:
   - RemoteSocket cannot have listeners attached
   - RemoteSocket ACKs behave differently (timeout is required, see https://github.com/socketio/socket.io/pull/4558)
   - CheatSheet for RemoteSocket (?)
- Add quick ways to distinguish between Socket and RemoteSocket (some `isRemote` flag or similar)

I've opened also a discussion about this some time ago: https://github.com/socketio/socket.io/discussions/4807

**Reply by darrachequesne** [reply]
Thanks for the heads-up, I will add this to the documentation.

**Reply by ghost** [reply]
Thank you!
